rs/index/src/ivf/files/invalidated_ids.rs

```diff
@@ -1,122 +1,167 @@
 
-use anyhow::{anyhow, Result};
-use memmap2::MmapMut;
+use anyhow::Result;
 
 pub struct InvalidatedIdsStorage {
-    mmap: MmapMut,
-    num_points: usize,
+    base_directory: String,
+    backing_file_size: usize,
+    files: Vec<File>,
+    current_backing_id: i32,
+    current_offset: usize,
 }
 
-const BITS_IN_BYTE: usize = 8;
-
 impl InvalidatedIdsStorage {
-    pub fn new(base_directory: &str, num_points: usize) -> Result<Self> {
-        let file = OpenOptions::new()
-            .read(true)
-            .write(true)
-            .create(true)
-            .open(format!("{}/invalidated_ids", base_directory))?;
-
-        let bytes_needed = num_points.div_ceil(BITS_IN_BYTE);
-        file.set_len(bytes_needed as u64)?;
-
-        let mmap = unsafe { MmapMut::map_mut(&file)? };
-        Ok(Self { mmap, num_points })
-    }
-
-    pub fn invalidate(&mut self, point_id: u32) -> Result<()> {
-        let point_id = point_id as usize;
-        if point_id >= self.num_points {
-            return Err(anyhow!("ID out of bound"));
+    pub fn new(base_directory: &str, backing_file_size: usize) -> Self {
+        let bytes_per_invalidation = size_of::<u128>() + size_of::<u32>();
+        let rounded_backing_file_size =
+            backing_file_size / bytes_per_invalidation * bytes_per_invalidation;
+        Self {
+            base_directory: base_directory.to_string(),
+            backing_file_size: rounded_backing_file_size,
+            files: vec![],
+            current_backing_id: -1,
+            current_offset: rounded_backing_file_size,
         }
+    }
 
-        let byte_index = point_id / BITS_IN_BYTE;
-        let bit_index = point_id % BITS_IN_BYTE;
-
-        self.mmap[byte_index] |= 1 << bit_index;
+    fn new_backing_file(&mut self) -> Result<()> {
+        self.current_backing_id += 1;
+        let backing_file = OpenOptions::new().create(true).append(true).open(format!(
+            "{}/invalidated_ids.bin.{}",
+            self.base_directory, self.current_backing_id
+        ))?;
 
-        Ok(self.mmap.flush()?)
+        self.files.push(backing_file);
+        self.current_offset = 0;
+        Ok(())
     }
 
-    // TODO(tyb): consider having an unsafe version w/o bound check
-    // for performance reason
-    pub fn is_valid(&self, point_id: u32) -> Result<bool> {
-        let point_id = point_id as usize;
-        if point_id >= self.num_points {
-            return Err(anyhow!("ID out of bound"));
+    pub fn invalidate(&mut self, user_id: u128, point_id: u32) -> Result<()> {
+        if self.current_offset == self.backing_file_size {
+            self.new_backing_file()?;
         }
-        let byte_index = point_id / BITS_IN_BYTE;
-        let bit_index = point_id % BITS_IN_BYTE;
 
-        Ok((self.mmap[byte_index] & (1 << bit_index)) == 0)
+        let file = &mut self.files[self.current_backing_id as usize];
+        file.write_all(&user_id.to_le_bytes())?;
+        file.write_all(&point_id.to_le_bytes())?;
```

Comment:
2 options:
* make this write atomic. meaning create a buffer, then copy user_id and point_id to that buffer, than make a single write_all call, then sync_all
* make the first 4 bytes number of entries, then you don't have to worry about partial writes, similar to [this](https://github.com/hicder/muopdb/blob/e38439f25a9aef2d57c1faf6e0b2760f5ca8b2e1/rs/index/src/wal/file.rs#L16)

Reply:
or put a TODO and do it later :p